modules/sfp_threatjammer.py

```diff
@@ -0,0 +1,238 @@
+                "Navigate to https://threatjammer.com/keys",
+                "The API Keys are listed under 'API Key' column"
+            ],
+            'favIcon': "https://threatjammer.com/favicon.ico",
+            'logo': "https://threatjammer.com/threatjammer-risk-score.png",
+            'description': "Threat Jammer is a service to access high-quality threat intelligence"
+            " data from a variety of sources, and integrate it into their applications with the"
+            " sole purpose of detecting and blocking malicious activity."
+        }
+    }
+
+    opts = {
+        'api_key': "",
+        'api_hostname': "dublin.api.threatjammer.com",
+        'risk_score_min': 35,
+        'checkaffiliates': True,
+    }
+
+    optdescs = {
+        'api_key': "API key",
+        'api_hostname': "User API hostname",
+        'risk_score_min': "Minimum Threat Jammer risk score",
+        'checkaffiliates': "Apply checks to affiliates?",
+    }
+
+    results = None
+    errorState = False
+
+    def setup(self, sfc, userOpts=dict()):
+        self.sf = sfc
+        self.results = self.tempStorage()
+        self.errorState = False
+
+        for opt in list(userOpts.keys()):
+            self.opts[opt] = userOpts[opt]
+
+    def watchedEvents(self):
+        return [
+            "IP_ADDRESS",
+            "IPV6_ADDRESS",
+            "AFFILIATE_IPADDR",
+            "AFFILIATE_IPV6_ADDRESS",
+        ]
+
+    def producedEvents(self):
+        return [
+            "BLACKLISTED_IPADDR",
+            "BLACKLISTED_AFFILIATE_IPADDR",
+            "MALICIOUS_IPADDR",
+            "MALICIOUS_AFFILIATE_IPADDR",
+        ]
+
+    def queryIp(self, ip):
+        """Query API for an IPv4 or IPv6 address.
+
+        Args:
+            ip (str): IP address
+
+        Returns:
+            str: API response as JSON
+        """
+
+        headers = {
+            'Authorization': f"Bearer {self.opts['api_key']}",
+            'Accept': 'application/json',
+        }
+
+        res = self.sf.fetchUrl(
+            f"https://{self.opts['api_hostname']}/v1/assess/ip/{ip}",
+            timeout=self.opts['_fetchtimeout'],
+            useragent=self.opts['_useragent'],
+            headers=headers
+        )
+
+        time.sleep(1)
+
+        if res['code'] == '429':
+            self.error("You are being rate-limited by ThreatJammer.com")
+            self.errorState = True
+            return None
+
+        if res['code'] == '401':
+            self.error("You are not authorized by ThreatJammer.com. Check your API key.")
+            self.errorState = True
+            return None
+
+        if res['code'] == '400':
+            self.error("ThreatJammer.com rejected the IP address. Use only public IP addresses.")
+            self.errorState = True
+            return None
+
+        if res['code'] == '422':
+            self.error("ThreatJammer.com could not process the IP address. Check the format.")
+            self.errorState = True
```

Comment:
Note that `errorState` should only be used in cases where the module should no longer be invoked for other events. It's used in cases where the upstream or module should no longer be called due to some kind of critical failure.